### PR TITLE
[Issue-3169] Added S3 Wrapper

### LINKED ISSUE #3169: Create script for downloading folder from S3 bucket

### PR DESCRIPTION
**Pull request by carloscaldas**
Closes #3169 

This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3171)

### CODE REVIEW

File: src/main/scala/scripts/s3wrapper/S3Wrapper.scala [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+  private val s3Client: AmazonS3 = AmazonS3ClientBuilder
+    .standard()
+    .withCredentials(new AWSStaticCredentialsProvider(awsCredentials))
+    .build()
+
+  private lazy val cachedBuckets = new AtomicReference[Set[S3Bucket]](listBucketNamesInternal)
+
+  /**
+    * Retrieve all available buckets available
```
Comment (nikolayilyin) [inline src/main/scala/scripts/s3wrapper/S3Wrapper.scala:35]: two same words

File: src/main/scala/scripts/s3wrapper/S3RemoteResource.scala [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import scala.annotation.tailrec
+
+sealed trait S3RemoteResource {
+  def key: String
+  def sizeInBytes: Long
+}
+
+case class S3RemoteFile(key: String, sizeInBytes: Long) extends S3RemoteResource
+
+case class S3RemoteDirectory(key: String, children: Set[S3RemoteResource] = Set.empty) extends S3RemoteResource {
```
Comment (dimaopen) [inline src/main/scala/scripts/s3wrapper/S3RemoteResource.scala:12]: I wasn't able to find usage of S3RemoteDirectory or S3RemoteResource. Does that mean that we can leave only S3RemoteFile?

File: src/main/scala/scripts/s3wrapper/S3Wrapper.scala [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    * shutdown method states "shuts down this client object, releasing any resources that might be held open."
+    */
+  override def close(): Unit = {
+    s3Client.shutdown()
+  }
+}
+
+object S3Wrapper extends StrictLogging {
+
+  val directories = Map.empty[String, S3RemoteDirectory]
```
Comment (dimaopen) [inline src/main/scala/scripts/s3wrapper/S3Wrapper.scala:146]: Is it used anywhere?

File: src/main/scala/scripts/s3wrapper/S3Wrapper.scala [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    * @param bucketName the bucket name
+    * @param keysStartingWith the object key is composed its full path
+    * @param destinationPath the directory which downloaded data will be saved
+    */
+  final def download(bucketName: String, keysStartingWith: String, destinationPath: Path): Unit = {
+    if (!Files.isDirectory(destinationPath)) {
+      throw new IllegalArgumentException(s"Destination path [$destinationPath] must be a directory")
+    }
+    if (Files.exists(destinationPath.resolve(bucketName))) {
+      throw new IllegalArgumentException(s"Destination path [$destinationPath] cannot contain a directory with the bucket name [$bucketName]")
```
Comment (nikolayilyin) [inline src/main/scala/scripts/s3wrapper/S3Wrapper.scala:122]: I propose different text for exception: 
Destination path [$destinationPath] already contains a directory with the name [$bucketName]''

**Review by nikolayilyin — APPROVED** [review]
a couple of nitpicks, the rest LGTM

File: src/main/scala/scripts/s3wrapper/S3RemoteResource.scala (hunk 2 above)
Comment (carloscaldas) [reply src/main/scala/scripts/s3wrapper/S3RemoteResource.scala:12]: There were some other code that was removed for this PR. so, you are right. I am removing it.

**Comment by carloscaldas** [comment]
test!

**Comment by carloscaldas** [comment]
test now!

**Comment by carloscaldas** [comment]
test!

**Comment by carloscaldas** [comment]
test now!

**Comment by carloscaldas** [comment]
Test now!

**Comment by carloscaldas** [comment]
Test now!

File: src/test/scala/scripts/s3wrapper/S3WrapperTest.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
+/* package scripts.s3wrapper
```
Comment (alex-vv) [inline src/test/scala/scripts/s3wrapper/S3WrapperTest.scala:1]: I forgot to uncomment this one

Comment (carloscaldas) [reply src/test/scala/scripts/s3wrapper/S3WrapperTest.scala:1]: the main problem is solved. thx. I will handle from now on

File: build.gradle [hunk 6]
```diff
@@ -11,8 +11,8 @@ buildscript {
         gradlePluginPortal()
     }
     dependencies {
-        classpath group: 'kr.motd.gradle', name: 'sphinx-gradle-plugin', version: '1.0.3.Final' // can be upgraded to 2.2.0
-        classpath "jp.classmethod.aws:gradle-aws-plugin:0.35"  // can be upgraded to 0.37
+        classpath group: 'kr.motd.gradle', name: 'sphinx-gradle-plugin', version: '1.0.3.Final'
+        classpath "jp.classmethod.aws:gradle-aws-plugin:0.41"
```
Comment (JustinPihony) [inline build.gradle:15]: We should use the same version across build gradles - 78229ee

`classpath "jp.classmethod.aws.reboot:gradle-aws-plugin-reboot:0.42"`

Sorry for the late reply on this one, also

**Pull request merged**